src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ReliableRequestSessionChannel.cs

```diff
@@ -0,0 +1,842 @@
+                {
+                    if (!_completed)
+                    {
+                        _aborted = true;
+                        _completed = true;
+                        _tcs?.TrySetResult(null);
+                    }
+                }
+            }
+
+            public void Fault(RequestChannel channel)
+            {
+                lock (ThisLock)
+                {
+                    if (!_completed)
+                    {
+                        _faulted = true;
+                        _completed = true;
+                        _tcs?.TrySetResult(null);
+                    }
+                }
+            }
+
+            public void Complete()
+            {
+            }
+
+            public async Task SendRequestAsync(Message message, TimeoutHelper timeoutHelper)
+            {
+                _originalTimeout = timeoutHelper.OriginalTimeout;
+                if (!await _parent.connection.AddMessageAsync(message, timeoutHelper.RemainingTime(), this))
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(_parent.GetInvalidAddException());
+            }
+
+            public void Set(Message reply)
+            {
+                lock (ThisLock)
+                {
+                    if (!_completed)
+                    {
+                        _reply = reply;
+                        _completed = true;
+                        _tcs?.TrySetResult(null);
+                        return;
+                    }
+                }
+                if (reply != null)
+                {
+                    reply.Close();
+                }
+            }
+
+            public async Task<Message> ReceiveReplyAsync(TimeoutHelper timeoutHelper)
+            {
+                bool throwing = true;
+
+                try
+                {
+                    bool expired = false;
+
+                    if (!_completed)
+                    {
+                        bool wait = false;
+
+                        lock (ThisLock)
+                        {
+                            if (!_completed)
+                            {
+                                wait = true;
+                                _tcs = new TaskCompletionSource<object>();
+                            }
+                        }
+
+                        if (wait)
+                        {
+                            expired = !await TaskHelpers.AwaitWithTimeout(_tcs.Task, timeoutHelper.RemainingTime());
+
+                            lock (ThisLock)
+                            {
+                                if (!_completed)
+                                {
+                                    _completed = true;
+                                }
+                                else
+                                {
+                                    expired = false;
+                                }
+                            }
+
+                            _tcs.TrySetResult(null);
```

Comment:
I don't quite understand why you're calling TrySetResult here. You've already awaited the _tcs above, in AwaitWithTimeout. So why call TrySetResult here?

Reply:
For some reason I thought that the tcs was await'ed somewhere else. I checked the code and that's not the case so this isn't needed. That now invalidates my prior comment about TrySetResult as it looks like it's protected by a lock and the _completed field.